### PR TITLE
Prevent historical state from being pushed to an application service via `/transactions` (MSC2716)

### PR DESCRIPTION
**Pull request by MadLittleMods**
Prevent historical state from being pushed to an application service via `/transactions`. We now mark historical state from the MSC2716 `/batch_send` endpoint as `historical` which makes it `backfilled` and have a negative `stream_ordering` so it doesn't get queried by `/transactions`.

Fix https://github.com/matrix-org/synapse/issues/11241

Complement tests: https://github.com/matrix-org/complement/pull/221

Part of MSC2716: https://github.com/matrix-org/matrix-doc/pull/2716

### Dev notes

```
update_membership
update_membership_locked
_local_membership_update
create_event

create_and_send_nonmember_event
create_event
```

`/sync` looks for `stream_ordering` but excludes all `outliers`.

https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/storage/databases/main/stream.py#L519-L527

Whereas `/transactions` only cares about `stream_ordering`. Perhaps we should exclude `outliers` in `/transactions`?

https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/storage/databases/main/appservice.py#L358-L368

 
## `/sync` stack trace

 - [`SyncRestServlet.on_GET`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/rest/client/sync.py#L113)
 - [`wait_for_sync_for_user`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L299)
 - [`_wait_for_sync_for_user`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L329)
 - [`current_sync_for_user`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L399)
 - [`generate_sync_result`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L1039)
 - [`_generate_sync_entry_for_rooms`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L1455)
 - [`_get_rooms_changed`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L1614)
 - [`get_room_events_stream_for_rooms`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/storage/databases/main/stream.py#L409)
 - [`get_room_events_stream_for_room`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/storage/databases/main/stream.py#L478-L527)

## `/transactions` stack trace

 - [`notify_interested_services`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/appservice.py#L66-L87)
 - [`_notify_interested_services`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/appservice.py#L90-L135)
    - [`get_new_events_for_appservice`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/storage/databases/main/appservice.py#L353-L368)
    - [`submit_event_for_as`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/appservice.py#L133-L135)

[`submit_event_for_as`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/scheduler.py#L94-L95) ->

 - [`enqueue_event`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/scheduler.py#L129-L131)
 - [`_send_request`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/scheduler.py#L137-L157)
 - `appservice.scheduler.send`
 - [`create_appservice_txn`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/scheduler.py#L196-L201)
 - [`AppServiceTransaction.send`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/__init__.py#L338)
 - [`push_bulk`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/api.py#L202)
 - [`/transactions`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/api.py#L220)
 
 

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/appservice/api.py
Comment (MadLittleMods) [inline synapse/appservice/api.py:253]: Without these extra params, it's pretty much impossible to debug errors like:

```
push_bulk to http://host.docker.internal:9111/transactions/1 threw exception 400 Bad Request

push_bulk to http://localhost:9000/transactions/1 threw exception An error occurred while connecting: 99: Cannot assign requested address.

push_bulk to http://0.0.0.0:9111/transactions/1 threw exception Connection was refused by other side: 111: Connection refused.

push_bulk to http://host.docker.internal:9111/transactions/1 received 504
```

When there is more info available like:
```
push_bulk to http://host.docker.internal:9111/transactions/1 threw PartialDownloadError code=b'400' msg=b'Bad Request' response=b'Client sent an HTTP request to an HTTPS server.\n'

push_bulk to http://localhost:9111/transactions/1 threw exception(ConnectError) An error occurred while connecting: 99: Cannot assign requested address. args=('Cannot assign requested address',)

push_bulk to https://host.docker.internal:9111/transactions/1 threw exception(RequestTransmissionFailed) [<twisted.python.failure.Failure OpenSSL.SSL.Error: [('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')]>]
```

File: synapse/handlers/room_batch.py
Comment (MadLittleMods) [inline synapse/handlers/room_batch.py:224]: All of these changes are just piping `historical` down to `create_event`

```
update_membership
update_membership_locked
_local_membership_update
create_event
```

File: synapse/handlers/room_batch.py
Comment (MadLittleMods) [inline synapse/handlers/room_batch.py:244]: All of these changes are just piping `historical` down to `create_event`

```
create_and_send_nonmember_event
create_event
```

File: synapse/appservice/api.py
Comment (erikjohnston) [reply synapse/appservice/api.py:253]: An alternate approach here is to do:

```python
logger.warning(
    "push_bulk to %s threw exception %s", 
    uri, 
    ex,
    exc_info=logger.isEnabledFor(logging.DEBUG),
)
```

which should include exception info if the logging level is set to debug. Your method also works

File: synapse/appservice/api.py
Comment (erikjohnston) [inline synapse/appservice/api.py:238]:
```suggestion
            if logger.isEnabledFor(logging.DEBUG):
                logger.debug(
                    "push_bulk to %s succeeded! events=%s",
                    uri,
                    [event.get("event_id") for event in events],
                )
```

Otherwise we'll always generate the list, even if the logging isn't set to debug

File: changelog.d/11265.bugfix
Comment (erikjohnston) [inline changelog.d/11265.bugfix:1]: I'm a bit confused why these would come down `/transactions` and not down `/sync`?

Comment (MadLittleMods) [reply changelog.d/11265.bugfix:1]: I was curious a little bit about this as well and saw that `/transactions` only cares about `stream_ordering` so it made sense why it still came down and didn't look into it further.

Diving into this more, here are the details:

`/sync` looks for `stream_ordering` but excludes all `outliers`.

https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/storage/databases/main/stream.py#L519-L527

Whereas `/transactions` only cares about `stream_ordering`. Perhaps we should exclude `outliers` in `/transactions`?

https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/storage/databases/main/appservice.py#L358-L368



 
## `/sync` stack trace

 - [`SyncRestServlet.on_GET`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/rest/client/sync.py#L113)
 - [`wait_for_sync_for_user`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L299)
 - [`_wait_for_sync_for_user`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L329)
 - [`current_sync_for_user`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L399)
 - [`generate_sync_result`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L1039)
 - [`_generate_sync_entry_for_rooms`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L1455)
 - [`_get_rooms_changed`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/sync.py#L1614)
 - [`get_room_events_stream_for_rooms`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/storage/databases/main/stream.py#L409)
 - [`get_room_events_stream_for_room`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/storage/databases/main/stream.py#L478-L527)


## `/transactions` stack trace

 - [`notify_interested_services`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/appservice.py#L66-L87)
 - [`_notify_interested_services`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/appservice.py#L90-L135)
    - [`get_new_events_for_appservice`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/storage/databases/main/appservice.py#L353-L368)
    - [`submit_event_for_as`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/appservice.py#L133-L135)

[`submit_event_for_as`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/scheduler.py#L94-L95) ->

 - [`enqueue_event`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/scheduler.py#L129-L131)
 - [`_send_request`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/scheduler.py#L137-L157)
 - `appservice.scheduler.send`
 - [`create_appservice_txn`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/scheduler.py#L196-L201)
 - [`AppServiceTransaction.send`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/__init__.py#L338)
 - [`push_bulk`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/api.py#L202)
 - [`/transactions`](https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/appservice/api.py#L220)

File: synapse/appservice/api.py
Comment (MadLittleMods) [reply synapse/appservice/api.py:253]: Maybe both 👀. The warnings are kinda useless without the little extra info on type of error and arguments. But also useful to have the stack when debugging.

examples with `exc_info=True`
```python
push_bulk to http://host.docker.internal:65302/transactions/1 received code=500 msg=Internal Server Error
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/appservice/api.py", line 229, in push_bulk
    await self.put_json(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/client.py", line 639, in put_json
    raise HttpResponseException(
synapse.api.errors.HttpResponseException: 500: Internal Server Error
```

```python
push_bulk to http://host.docker.internal:49443/transactions/1 threw exception(ConnectionRefusedError) Connection was refused by other side: 111: Connection refused. args=('Connection refused',)
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/appservice/api.py", line 229, in push_bulk
    await self.put_json(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/client.py", line 630, in put_json
    response = await self.request(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/client.py", line 440, in request
    response = await make_deferred_yieldable(request_deferred)
twisted.internet.error.ConnectionRefusedError: Connection was refused by other side: 111: Connection refused.
```

File: changelog.d/11265.bugfix
Comment (richvdh) [reply changelog.d/11265.bugfix:1]: Apologies for jumping in without context, but

> Perhaps we should exclude outliers in /transactions?

the outlier flag seems like a poor way to decide whether we should push this data. (Yes, outliers shouldn't be sent over `/transactions`, but there are probably many other events which shouldn't be sent).

`FederationEventHandler._process_received_pdu` has a `backfilled` parameter (see https://github.com/matrix-org/synapse/blob/b09d90cac9179f84024d4cb3ab4574480c1fd1df/synapse/handlers/federation_event.py#L945), whose purpose is slightly unclear, but I *think* one of its jobs is this sort of thing. Maybe we should use similar logic to that, somehow?

File: synapse/handlers/room_batch.py
Comment (richvdh) [reply synapse/handlers/room_batch.py:244]: whatever else happens, please please please make sure that any parameters are added to the docstrings *at each level*, with the intended purpose clearly described. 

Relatedly: as a general rule, I would say it is preferable for method parameters to describe a change in behaviour ("`inhibit_sync_to_clent`") rather than describe some bit of state that the function is free to interpret as it wishes: it's much clearer for a later reader to understand how things tie together, particularly when things then get modified later on. Obviously this isn't a hard-and-fast rule, but it's worth considering.

[Currently we have a bit of a mess in some places (`FederationEventHander.backfilled` being a prime example) where you have to dig through 150 methods to find out what a flag actually does, and then somehow reverse-engineer what it is *supposed* to do. Don't do that.]

Comment (MadLittleMods) [reply synapse/handlers/room_batch.py:244]: > please please please make sure that any parameters are added to the docstrings at each level, with the intended purpose clearly described.

Will do 🙇 

---

> Relatedly: as a general rule, I would say it is preferable for method parameters to describe a change in behaviour ("inhibit_sync_to_clent") rather than describe some bit of state that the function is free to interpret as it wishes

Seems reasonable. I've created a separate issue to track this https://github.com/matrix-org/synapse/issues/11300

File: changelog.d/11265.bugfix
Comment (MadLittleMods) [reply changelog.d/11265.bugfix:1]: Whatever we decide on to make `/sync` and `/transactions` behave similarly, we can tackle in another PR. This PR is still good to ship on its own, i.e. It's a good idea to mark historical state as `historical`

---

> `FederationEventHandler._process_received_pdu` has a `backfilled` parameter [...]
> Maybe we should use similar logic to that, somehow?

For `/transactions`, we can't tell whether the event was `backfilled`. The only indication is that the `stream_ordering` would be negative which is what the `/transactions` code already takes into account. And is why the fix in this PR to mark the historical state events `historical` -> `backfilled` works to not show up in `/transactions`.

I just suggested adding `outlier` to `/transactions` query as well so it would match the query in `/sync`

File: synapse/appservice/api.py
Comment (erikjohnston) [reply synapse/appservice/api.py:259]: Fair. Ideally converting an exception to a string should give us useful information, but clearly those exceptions (which we don't control) don't follow that rule :/

**Comment by erikjohnston** [comment]
This approach seems reasonable to me as a stop gap before we try and use the same rules for sending events down `/sync` and `/transactions`

**Comment by MadLittleMods** [comment]
@erikjohnston Approval please 🙂 

Do you want @richvdh to approve as well before merge?

**Comment by richvdh** [comment]
given erik has the context here, I'm happy to defer to him, if he has the bandwidth to look at it. I was just making driveby comments before!

**Review by erikjohnston — APPROVED** [review]

**Comment by MadLittleMods** [comment]
Thanks for the review @erikjohnston and @richvdh 🐖

File: changelog.d/11265.bugfix
Comment (MadLittleMods) [reply changelog.d/11265.bugfix:1]: Created https://github.com/matrix-org/synapse/issues/11394 to track this ⏩

File: synapse/appservice/api.py
Comment (MadLittleMods) [inline synapse/appservice/api.py:234]: I came back to this and was confused why I did this since it's guarding debug against debug.

AFAICT, this is to avoid potentially expensive string construction when we loop through all of the `events` here. We do this in a few other places in the codebase too.

**Pull request merged**
